Cfamutator exceptions

o-petrov opened the pull request.

o-petrov committed:
Set CFA creator statistics out of CPAchecker#parse

So this method does not need MainCPAStatistics as a parameter.

o-petrov committed:
Make inner class from CPAchecker#run

It allows to separate error handling from main flow,
which is useful for the following commits.

o-petrov committed:
Catch parsing exceptions in CPAchecker#parse instead of #run

o-petrov committed:
Expand CPAchecker$Mutator stub

Set CFA creator stats out of #setup, as intermediate
analysis runs should not have it.

Add AnalysisResult triple, as comparing reached set,
CFA, or statistics does not make sense.

o-petrov committed:
Split analysis setup and run

To move CFA creation (and stats) out of setup.

o-petrov committed:
Add stub for CPAchecker#Mutator

It will run analysis, check for expected result,
and mutate CFA while possible.

o-petrov committed:
Split CPAchecker#parse into parse and load from serialized CFA file

o-petrov committed:
Pass CFACreator as argument to CPAchecker#parse

CFAMutator needs info about functions' local CFA before
any processings in CFACreator#createCFA, so it will
extend CFACreator.

o-petrov committed:
Replace ParseResult with new class FunctionCFAsWithMetadata

It extends ParseResult and stores CFA of each function and list of
global declarations. It stores not only nodes, but local edges too.

Also it stores language, machine model, and main function like
MutableCFA does. The main function is needed even for proper CFA
creation.

o-petrov committed:
Change strategy stub to fit Delta Debugging approach

o-petrov committed:
Add generic Delta Debugging stategy and node-edge removing stub

The 'stub' should be usable already and is capable to remove nodes
that have single leaving edge.

o-petrov committed:
Fix small bugs

o-petrov committed:
Add logging to GenericDD strategy

o-petrov committed:
Fix bug: end before halving in empty delta list

o-petrov committed:
Change Local CFA edge collection

As order of edges must be stored too, the simpliest way is to
store 'nodes' with correct local edges in correct order.

Save after original ParseResult and every mutation, reset before
first #canMutate and every #setResult.

o-petrov committed:
Fix exceptions comparason

They were always different, so it always resulted in UNRESOLVED.

o-petrov committed:
Change rollbackInfos from Stream to ImmutableList

We need mutations to be applied immediately in #mutate,
but stream is lazy, so it would be done sometimes in
#setResult.

o-petrov committed:
Use CFAMutationUtils for more readability

o-petrov committed:
Fix bug: AssumeEdges copy-pasted

o-petrov committed:
Print edges of a problematic node

o-petrov committed:
Add CFAMutationStrategy javadoc

o-petrov committed:
Add function removing strategy

The strategy removes function nodes and function entry from CFA.
Main function is ignored for now (could be replaced with single
return statement).

o-petrov committed:
Add composite strategy, that uses sequence of strategies

o-petrov committed:
Fix DD finish

Allow empty delta list, as we have to check unresolved
objects a bit later.

Also fix two log strings.

o-petrov committed:
Move CFA export files to one subdirectory

o-petrov committed:
Reset information in CFANode

o-petrov committed:
Add description what does the strategy remove

o-petrov committed:
Refactor SingleEdgeRemover

Store nodes instead of edges as rollback information, as nodes
can be reconnected after info is stored.

Do not include nodes with branching predecessors: if both
successors of such a node are removed, there can be parallel
edges.

o-petrov committed:
Add strategy that removes statement and blank edges only

o-petrov committed:
Add strategies to remove chains and some cases for branching

If one of AssumeEdges is going a node with other
entering edges, it is possible to remove this
AssumeEdge.

o-petrov committed:
Add logging to CompositeStrategy

o-petrov committed:
Add safe objects removing to GenericDDStrategy

It would be nice to remove all safe objects, but sometimes
it can not be done for some reason.

o-petrov committed:
Export CFA after every analysis run and after every rollback

o-petrov committed:
Create new branch for CFA mutator

The goal is to mutate input CFA in such a way, that the reuslt of
analysis -- specifically, an exception thrown during analysis --
remains the same. This way a big input program which crashes
CPAchecker can be minimized automatically.

The work is based on Delta Debugging algorithm. This is refactoring
of https://github.com/mutilin/cpachecker/tree/cfamutation.

o-petrov committed:
CFAMutator extends CFACreator properly

Add new package mutation for strategies and move CFAMutator there.

o-petrov committed:
Bugfix

Empty deltas are possible in the last strategy round

o-petrov committed:
Add parameter to branching remover

Try to remove right and left assume edges in separate
passes, so if both can be tried, both will be tried.
Before only one was tried, and other was not considered.

o-petrov committed:
Export CFA before analysis run, not after

E.g. termination analysis changes CFA, so it can not be exported
to C. So export before analysis and after rollback.

o-petrov committed:
Output CFA even with -noout

o-petrov committed:
Do not nullize CFA if it was not rollbacked

It may be used in output.

o-petrov committed:
Add a few tests from gitlab issues

o-petrov committed:
Separate ldv-related programs into another benchmark file

They need a lot of time (how much exactly?)

o-petrov committed:
Add strategy to remove empty branches

Also move ChainRemover#isInsideChain to CFAMutationUtils
and RollbackInfo for SimpleBranchingRemover inside that class.

o-petrov committed:
Extract reconnecting entering edges to two new CFAMutationUtils methods

o-petrov committed:
Change branch detection in EmptyBranchPruner

Branches don't have to be chains. Chains that don't have common
successor may be full branches only if end at different sinks.

o-petrov committed:
Add stats to CFA mutator

o-petrov committed:
Add tests from issue 992

Move .xml into new subdir.
Comment 779 as it timeouts, 778a as it is parse
error, not analysis.

o-petrov committed:
Add statistics to strategies and CPAchecker

o-petrov committed:
Fix bug: no round needed to remove 0 safe objects

o-petrov committed:
Fix bug: collect stats from mutator after runs

Composite strategy adds stats only from used strategies,
and no strategies are used at start.

o-petrov committed:
Use loggers with strategy names, so they are not all GDD

o-petrov committed:
Fix smg-ldv tests

o-petrov committed:
Collect mutator statistics even if interrupted

o-petrov committed:
Deal with exceptions with empty stack trace

It seems to be JVM optimization for the case of recurring error.
Such an exception has no message to, so log only its class.

Also fix throwable comparing: messages can be formatted with
different information anyway, so check for exception class, and
first stack trace element if present.

o-petrov committed:
Fix bug: dont fail assert on branches of single AssumeEdge

But such edges can not be removed by EmptyBranchPruner,
as they would be removed by SimpleBranchingRemover already.

o-petrov committed:
Refactor statistics

o-petrov committed:
Fix ERROR title in benchmark tables

o-petrov committed:
Add two log lines between analysis and rollback

o-petrov committed:
Fix export

Do not ignore noout in CFACreator and do not
allow -noout in CFAMutator.

Add 0 to make original-cfa subdir in line with others.

o-petrov committed:
Add strategy that removes chains of statements only

* statement or blank edges.

Rename StatementRemover to StatementEdgeRemover.
Change constraint on node-edges that can be removed
by SingleEdgeRemover.
Change constraint on node-edges for StatementEdgeRemover:
do not try to remove single statements/blanks that must
be tried by StatementChainRemover before as short chain
of one edge.

o-petrov committed:
Add config option for debugging

o-petrov committed:
Allow UNKNOWN without exception and

Do not replace DONE with last result of analysis
because of problems with witness export.

o-petrov committed:
Add three strategies (expr & decls)

1. Remove expressions by replacing them with some default literal.
2. Remove local variables declarations if they are not used.
3. Remove global declarations of unused variables, types, functions.

o-petrov committed:
Add reset to declaration strategies

They wont make next passes otherwise.
Also dont try to remove declarations already found
to be neccessary in previous passes.

o-petrov committed:
Add, not replace global declaration by index

o-petrov committed:
Do not add last pass to counter

Last time #canMutate returned false, so there
was no mutation round.

o-petrov committed:
Move sinks collection to CFAUtils

o-petrov committed:
Factor Mutator class from CPAchecker into CPAcheckerMutator

Unextract CPAchecker$Analyzer back to CPAchecker, and move $Mutator
to new class CPAcheckerMutator. There were too many inner classes
and too little encapsulation.

PavelAndrianov commented on src/org/sosy_lab/cpachecker/cfa/export/DOTBuilder2.java:
```diff
@@ -236,7 +236,7 @@ private void postProcessing() {
 
     void writeFunctionFile(String funcname, Path outdir) throws IOException {
 
-      Path cfaFile = outdir.resolve("cfa__" + funcname + ".dot");
```
Зачем такие изменения?

o-petrov replied:
вообще касательно экспорта cfa достаточно было только добавить папку куда кладётся всё что касается cfa, но я решил переименовать опции и, вот, экспортировать cfa функций под их именами без префикса.

PavelAndrianov replied:
Я все равно не понял, зачем менять общий вывод для всего cpachecker. Сделайте себе опцию.

o-petrov replied:
Отменил/переделал всё что касается опций экспорта

PavelAndrianov commented on src/org/sosy_lab/cpachecker/cfa/model/CFANode.java:
```diff
@@ -328,4 +329,108 @@ public Set<CSimpleDeclaration> getOutOfScopeVariables() {
+   * Insert new leaving edge instead of old one. Used only for consistent CFA mutation. Warning:
+   * everything is checked for equality by ==, not equals, as there should not be any copies.
+   */
+  @Deprecated
```
Почему они все Deprecated?

o-petrov replied:
это хак, чтобы показать, что метод никто (кроме меня) использовать не должен (в C++ на такие случаи вроде есть друзья класса или типа того). так же, например, поступили с методами где-то в MachineModel, которые расчитывают всякие сдвиги в структурах, и используются в SMG.

PavelAndrianov commented on src/org/sosy_lab/cpachecker/cfa/mutation/AbstractExpressionSubstitutor.java:
```diff
@@ -0,0 +1,359 @@
+import org.sosy_lab.cpachecker.cfa.types.java.JType;
+import org.sosy_lab.cpachecker.util.Pair;
+
+abstract class AbstractExpressionSubstitutor {
```
Хорошо бы иметь краткое описание класса: зачем он и что делает.

o-petrov replied:
добавил по небольшому комментарию к Abstract- и ToDefaultExpressionSubstitutor

PavelAndrianov commented on src/org/sosy_lab/cpachecker/cfa/mutation/CFAMutator.java:
```diff
@@ -0,0 +1,197 @@
+    }
+    cfaExportDirectory = exportDirectory;
+    List<CFAMutationStrategy> subs;
+    if (order == null) {
```
Нельзя ли сделать нормальное перечисление элементов, например, через enum? Сокращение по первым буквам нечитаемо.

o-petrov replied:
изначально я надеялся обойтись без настройки мутатора, но так как настройка полезна как минимум для экспериментов и остаётся надолго, то я исправлю это на список классов.

o-petrov replied:
исправил

PavelAndrianov commented on src/org/sosy_lab/cpachecker/cmdline/CPAMain.java:
```diff
@@ -746,10 +759,10 @@ private static void printResultAndStatistics(
-            makePrintStream(
-                mergeStreams(System.out, file)); // ensure that result is printed to System.out
+        // ensure that result is printed to System.out
+        stream = makePrintStream(mergeStreams(System.out, file));
```
Необязательное изменение

o-petrov committed:
Add more CFA checks

Check full CFA by checking all function CFAs.
Add checks for summary and interprocedure edges.

o-petrov committed:
Improve option for CFA mutation strategy

Use list of classes instead of string of letters.
Every strategy thus must have public constructor
with single LogManager parameter.

o-petrov committed:
Use new CPAchecker instances for each analysis round

Use specified time limit for a round.
TODO use specified logger for a round.

o-petrov committed:
Allow inconsistent full CFA.

Because of #989 full CFA can be inconsistent, so warn and continue.

o-petrov committed:
Refactor CPAcheckerMutator a bit

Extract new method to produce result of mutations same way for
every return. Store CPAcheckerResult (instead of just verdict and
description) as reached set, cfa, stats are needed for exports
in CPAMain after CPAcheckerMutator return.

Also move time measuring and its logging inside #analysisRound.

o-petrov committed:
Add check whether there is enough time for next analysis run

If there is not, interrupt now. Avoid forced shutdowns because of
analysis longer than ForcedShutdown limit for hanging.

Add options to manipulate walltime limit for rounds.

o-petrov committed:
Reword messages in CFAcheck

o-petrov committed:
Collect not only direct types

It was a bug for GlobalDeclarationRemover.

o-petrov committed:
Improve AssumeEdge simple case strategy

Rename SimpleBranchingRemover to AssumeEdgesRemover.

Instead of two instances for two sides use one in two passes
(declaration removers work in several passes too).

o-petrov committed:
Enhance expression substitutor

Better handle initializers, handle function call assignments,
add javadocs.

o-petrov committed:
Change option for checks after rollback

If there is a problem with mutations, such that rollback can not
restore dought-for bug occurance, every following round will be
rollbacked too. So instead of checking every rollback, only
suspiciously many rollbacks in a row must be checked. On the other
hand, we want to find it as soon as possible.

So make option for how many rollbacks in a row are suspicious.

o-petrov committed:
Log analysis to file

Logging every round to console is too much,
but no logging at all seems to be bad too.

o-petrov committed:
Change log level for FunctionBodyRemover

o-petrov committed:
Remove nodes that are reachable via summary edges too

If a node is after summary edge of a non-returning function, it is
reachable only via that summary edge (there is no return edge). In
this case every node that is reachachable from this node only is
removed, but the node that is reachable via possible leaving summary
edge was not considered.

o-petrov committed:
Change benchmarks

Add tests known to fail from Klever.
Remove many uthash files from termination.xml as they are
the same two files.
Avoid CPU time limit, use walltime limit.

o-petrov committed:
Get reason of shutdown correctly

Shutdown can be requested because next analysis run
could be too long. In this case shutdownNotifier throws
if it is asked about the reason.

Construct the reason in this case.

o-petrov committed:
Improve GDD javadoc: explain ddmin better

o-petrov committed:
Revert renaming of CFA export options

Renaming was done in commit
'Move CFA export files to one subdirectory'.

CFA is still exported in its own subdirectory.
New option for function CFAs subdirectory remains.
New boolean option "whether to export as pixel graphic" remains, as
without it pixel graphic file is exported unless -noout is specified.

o-petrov committed:
Redo CFA export subdirectories

Make CFA export subdirectory '.' by default, function CFAs
subdirectory 'cfa' by default, and revert 'cfa__' prefix,
so output is same as was in trunk.

o-petrov committed:
Add 100s so CPAchecker can finish gracefully

Note option limits.time.wall=1000s

o-petrov committed:
Add Javadocs to strategies

o-petrov committed:
Yet another fix for CFA export config

Revert names for function calls dot- and pixel
graphic files.
Check if directory is null before exporting
(although it is checked in #export anyway).

o-petrov committed:
Uncomment tests commented by mistake

o-petrov committed:
If cannot remove function, replace its body with dummy return

In case of main function and functions called via thread creation
operations the body must be present in the CFA. To remove such a
function replace its body with single return.

For some types default expressions are harder to construct than
default initializer. In such case insert dummy variable to return.

o-petrov committed:
Fix config options file

o-petrov committed:
Check if any argument was actually substituted

o-petrov committed:
Do not mess thread operation arguments

o-petrov committed:
Do not replace chains of single blank edges

The strategy replaces a chain with a blank edge,
so it should not touch single blank edge chains.

o-petrov committed:
Fix function CFAs export option in benchmarks

o-petrov committed:
Less log limits

Move round's limits logging to the round log file.
Log setup limits once, as it does not change.

o-petrov committed:
Fix check for serialized CFA

o-petrov committed:
Experimental generalizations of DD

Actually my previous implementation, GenericDeltaDebuggingStrategy,
is DD cause-isolating algorithm plus some kind of minimization of
remaining safe part.

Implement DD cause-isolation and DD test-minimization algorithms,
plus the same as old implementation*, plus consequent DD CI.

*This one needs to be able to remove only deltas on second pass,
hence PartsToRemove enum.

Factor CFA manipulation out, as there are a few DD algorithms now.

If these are ok, implement Hierarchical DD.
It will be generalized for graphs, not only trees, hence
NeedsDependencies enum, which tells how to behave if a few
predecessors are present.

o-petrov committed:
Rename Delta Debugging algorithms

Previous names are long. New names are same as in papers.

o-petrov committed:
Rename two methods in DD

o-petrov committed:
Refactor abstract methods in DD. Add ddmax

o-petrov committed:
Add short descriptions to DDs

o-petrov committed:
Refactor logging

o-petrov committed:
Fix bug: ddmin can not reduce after PASS

o-petrov committed:
Rename other two based on DD algorithms

o-petrov committed:
Encapsulate some fields

o-petrov committed:
Refactor dradda: use two separate delegates

o-petrov committed:
Refactor DD constructors and setup

Use one constructor. Give elements to work with via #workOn method.
Retrieve all elements implicitly from element manipulator if they
were not provided.

DD algorithms can be used as mutation strategies, i.e. without
explicit setup with list of elements to work on.

DD algorithms can be used by other strategies to work with given
elements, if #workOn is called after constructor.

DD algorithms can be used by other strategies repeatedly without
constructing new instances, if #workOn is called repeatedly after
algorithm has ended on previous given elements.

o-petrov committed:
Collect multipass stats

o-petrov committed:
Allow DD algorithms from options

o-petrov committed:
Fix stats for dd* aka repeated dd

o-petrov committed:
Fix NPE: halve delta without referencing currentDelta

o-petrov committed:
Fix stats naming

o-petrov committed:
Fix bug for ONLY_COMPLEMENTS part to remove

In often case of two deltas, algorithm tries to remove the second
one as a complement. If test passes, the first delta is marked as
safe, and the second delta is restored. The bug is that the algorithm
tries to remove 'the first delta' while it is already marked as safe,
i.e. removes 0 elements.

For this case split deltas if iterator has no next delta or there
is only one delta remaining.

o-petrov committed:
Add edges to function call graph in function manipulator

o-petrov committed:
Rename class that represents an element of function (body) in a CFA

o-petrov committed:
Fix bug: do not add functions without CFA to the manipulator's graph

Edge's endpoints are added silently, but we do not need functions
which we can not remove or restore.

o-petrov committed:
Use new function-removing DD in big programs

o-petrov committed:
Fix: do not verify FAIL after rollback if mutations shutdown requested

o-petrov committed:
Print graph from DD statistics

If a hierarchy/dependency graph is added to DD stats, it is
printed to the output directory.

o-petrov committed:
REDO change manipulator

o-petrov committed:
Add DDs to benchmark

o-petrov committed:
REDO HDD

o-petrov committed:
NOT READY other HDD

o-petrov committed:
[TOO MUCH] September backlog

o-petrov committed:
Change walltime factor from a pair of integer nominator and denominator to one double

o-petrov committed:
Revert changing cil file

o-petrov committed:
Rewrite DD so it includes ddmin and ddmax, remove other variants

Remove all DD classes except for DD (now implements dd, ddmin and ddmax)
and HDD. Add outcomes to express properties of sought-for optimum/cause,
two properties (for minimization and for maximization) can be specified
(cause-isolation needs both).

o-petrov committed:
Fix bug: restore functions in call graph even if they are not called and do not call

o-petrov committed:
Add option to limit wall time for original run

o-petrov committed:
Fixes: comply to checkstyle

o-petrov committed:
Fix after-rollback-check: min-property should hold, even while maximazing

o-petrov committed:
Fix scope of CFAMutator dd-property methods

o-petrov committed:
DD remembers removed elements.

DD* (repeating DD) can naturally minimize by putting aside
already found cause and finding a new one among safe/max
elements. To use DD* for maximization it needs to use removed
elements as minimizing DD* uses safe elements.

o-petrov committed:
Rename back: max elements to safe, min elements to cause; small fixes

o-petrov committed:
Add CFAElementManipulator#restore to restore removed elements

o-petrov committed:
Add dd* algorithm

When dd resolves all elements, work on safe (in case dd*min)
or removed (in case of dd*max) elements to retrieve another
cause. Repeat while there are safe/removed elements after dd run.

o-petrov committed:
Fix bug: should not call #canMutate twice

o-petrov committed:
Small rewording

o-petrov committed:
make DD's private stage protected; fix bug in HDD

o-petrov committed:
Log dd*

o-petrov committed:
Add DDSearch, that iterates over optimums found by DD*

    DD* finds a minimum that consists of a several causes, or
    a maximum that consists of everything but a sequence of causes.
    DDSearch forces these causes out/in to find new min/max

o-petrov committed:
Cache dd results

Manipulators have to provide present objects for
algorithms other than plain dd.

o-petrov committed:
Move #shortListToLog to AbstrDDStr

o-petrov committed:
Change inner state in #setResult, not #canMutate

Also fix bug of double halving after REMOVE_WHOLE

o-petrov committed:
Change inner state in #testUnresolved

In other cases, #reduce and #increase, inner state was already
changed, so changing it generally in #setResult after that is a bug.

o-petrov committed:
Verify max-property when max finished

Min-property is checked sometimes after rollback.
But when maximization finishes, it also reports a rollback,
and also removes the cause, so maximization property
should hold instead.

o-petrov committed:
Fix bug: interrupt mutations between rounds if next analysis has little time

o-petrov committed:
Change inner state DD when removing a delta

o-petrov committed:
DD*Max should not remove whole

If the first found cause is the only cause, dd*max will remove it,
remaining CFA will PASS the test, so it will be found as maximum.

o-petrov committed:
Rename function call graph field in FunctionBodyManypulator

o-petrov committed:
Add element relation type parameter to CFAElementManipulator

o-petrov committed:
Refactor CFAElementManipulators in abstract class

Move common logic from FunctionBodyManipulator and
StructuredBlockManipulator.

Add method to restore once removed elements, hopfully
even if they were pruned.

(PS. The block manipulator is not used at all for now.)

o-petrov committed:
Fix bug: reset delta list with remaining half, not with just removed one

o-petrov committed:
Replace for-each with index for

The lists are growing so for-each can not continue on changed list

o-petrov committed:
Add CHECK_WHOLE stage so DD*Max checks whole before repeating

If the first found cause is the only cause, dd*max will remove it
in the first pass, return removed, and check all but the cause.
As test passes, dd*max found maximum.

If dd*max can not check whole, it assumes there are other causes,
so it will 'find' them and remove some elements that can safely
remain in maximum.

o-petrov committed:
Add test-set for timeout search

o-petrov committed:
Backup graph once, not after every #restore

o-petrov committed:
Don't forget to reset DD*. Rename list of causes

Fix bug: DDSearch must reset DDStar's lists before looking
for new optimum.

Rename cause list to list of causes, etc.

Fix bug: use index of list of causes, not iterator over
this list, as it is changing.

Change logging.

o-petrov committed:
Do not check rollback when #finalize was called

o-petrov committed:
DD*Min* should check if min-property still holds when a cause is removed

o-petrov committed:
Add a few assertions for delta list and unresolved elements

o-petrov committed:
Rename CFAMutator method to #clearCreatorStats

o-petrov committed:
Do not export coverage, as it fails

o-petrov committed:
Check feasibility of an error

If a search variant of DD is used, every FALSE verdict is
checked for feasibility. If a feasible error path is found,
search stops.

o-petrov committed:
Fix SpotBugs warnings

o-petrov committed:
Check if DD* can actually find a cause

If no cause is found, nothing more can be done.

Also deal with #CHECK_WHOLE in completly in DDStar,
as DD does not know which direction we are repeating it.

o-petrov committed:
Cover problem with witness export

If there are no edges between an ARG state and its child found,
a null edge was passed causing NPE. Pass dummy edge instead.

o-petrov committed:
Change time limit scheme for analysis rounds

I find hardcap + softcap more useful than hardcap original run
and then softcap other runs based on time consumed by it.

o-petrov committed:
Add config files. Add programs from SVCOMP22

Move common options from benchmark xml files to config properties
files.

Add a benchmark with some SVCOMP22 programs with timeout result.

o-petrov committed:
Fix bug: add edges for the node just added, not its predecessor

o-petrov committed:
Do not override #get...Elements methods, make new #getAll...Elements

o-petrov committed:
Log DD* better. Do not assume direction, just warn about it.

o-petrov committed:
Restore elements accordingly to how they were removed/pruned

o-petrov committed:
Print a bit more details

o-petrov committed:
Check feasibility on 'original' CFA

Restore edges and nodes as they were in the original CFA,
and remove them after feasibility check.

o-petrov committed:
Refactor DD*, DD**

Also do not store found optimums, as this is unused.
Also, when DD* returns no new cause, repeat it on new
forced cause with elements to resolve from last time.

o-petrov committed:
Comply to configuration checks

o-petrov committed:
If DD* finds empty maximum, do not continue DD**

o-petrov committed:
Add time limit for CFA exports

o-petrov committed:
Rearrange configs (comply with config checks)

1. Move cfa-mutation .properties files to test/ as they are used
only in test-sets.
2. Add short descriptions inside these test-sets.
3. Fix cfaMutation option description.

o-petrov committed:
Comply with SpotBugs

o-petrov committed:
Add FunctionBodyManipulator as default

o-petrov committed:
If analysis finishes correctly, just return the result

o-petrov committed:
Check feasibility of counterexamples instead of paths to targets

Previously feasibility of paths to target states was checked.
It seems to be tautological, so replace it with proper
CounterexampleCheckAlgorithm.

Add timelimit for this check, so it is not needed to be specified
differently for cex checker variants.

o-petrov committed:
Make possible to dump a file for each CEX checked by CBMC or concrete executor.

o-petrov committed:
Move time limit options from CPAcheckerMutator to a new class

o-petrov committed:
Refactor time limit factories to CFAMutationOptions methods

o-petrov committed:
Move shutdown check to CFAMutationLimits (prev. named Options)

o-petrov committed:
Rename CFA mutation limit options, descript them better

o-petrov committed:
Move round config, logger, etc setups to CFAMutationLimits
Setup a new directory for following analysis/CEX check,
produce correct config with explicitly specified limits,
then return CPAchecker/CEXcheckAlgorithm to CPAcheckerMutator.

Add option to choose CEX checker separately from main analysis.

Add a state to waitlist, so CounterexampleCheckAlgorithm checks
error paths. Add CLEAR_WAITLIST variant of NoopAlgorithm, so it
runs only once. TODO fix this hack.

o-petrov committed:
Rename three things:

class CFAMutationLimits -> CFAMutationManager,
enum ExportDirectory -> AnalysisRun (rename enumerators too),
method createLoggerAndStartLimitsForCheck -> createCEXCheckerAndStartLimits.

Remove debug prints.

o-petrov committed:
Remove CounterexampleCheckAlgorithm, use CounterexampleChecker directly

Removes hack around algorithm behavior (problem was, CEX check
algorithm starts (runs Noop algorithm and then does CEX check)
if reached set has waiting states).

Problem with export-to-C remains: missing some declaraions.

o-petrov committed:
Export from CEX checker to files numbered by CEX id

If there is no counterexample info present in the error state,
use 0 instead of throwing exception.

o-petrov committed:
Make try block smaller

o-petrov committed:
Factor CEX export from CounterexampleChecker#checkCounterexample

All three CEX checkers write CEX in their format to a file, then
check it. Separate these two things.

Throw CounterexampleAnalysisFailed instead of CPAException in
CexChecker#checkCounterexample.

Specify class explicitly for configuration injection.

o-petrov committed:
Do not use includes (and do not build config for that)

Also change return type of __VERIFIER_nondet_bool in exporter for
concrete executor.

o-petrov committed:
Append definiions of restored functions between CEX export and check

CounterexampleChecker exports CEX using mutated CFA. Append
restored functions from restored CFA to same file, and check it.

Do not restore CFA before usual CEX export.
Do not store original CFA. It can be retrieved from restoring it.
Do not store original CPA. It can be dummy, and ReachedSet has the
one that was used.
Do not add ARG CPA, it should be present anyway.

o-petrov committed:
Refactor CEX checkers, add CEX-as-C-program-CPAchecker

o-petrov committed:
Log exceptions of analysis round better

Do not rely on shutdownNotifier#getReason to get the message in
case of InterruptedException.

(ParallelAlgorithm#handleFutureResults throws it because of
ListenableFuture#get, not with shutdownNotifier.)

Log exception's stack with the round logger.

o-petrov committed:
Add factors to hack limits set in round configs

Time limit options are produced with respect to time limits
that are (will be) setup for the round. I approximate time that
round can spend as time to next check.

For wall time limit it is straightforward call
to #nanoSecondsToNextCheck(#getCurrentValue).

As N CPUs may use N times more time, CPU time limit may exceed in
1/N of given time, so it must be checked then. So multiply nanos
to check by N for CPU time limit.

(I didn't really check why 2 for Thread time limit.)

o-petrov committed:
Update configs and test-sets

Add cfaMutation-svcomp22 and cfaMutation-smg-ldv configs,
use them in respective test sets.

Remove DD and DD* rundefinitions from smg-error-search.xml
(ignore whitespace changes there)

o-petrov committed:
Dump stats every round

Dump intermediate statistics after every analysis round.
Dump frequency of CFANodes as locations of reached set states of result.
Do not add last round stats to the total stats.
Improve three logs.

o-petrov committed:
Add options to config in case of restart algorithm

It cannot run after it finds an error.
Do not print stats to main log every time a subalgorithm finishes.
Do print stats to a file every time whole algorithm finishes.

o-petrov committed:
Move single suppression from SpotBugs config to annotation

o-petrov committed:
Comply with all-checks small fixes

o-petrov committed:
Comply with configuration-check

Move options for restart algorithm to cfaMutation-svcomp22 config.
Exclude cpa.arg.logARGs, comment out CEX exports as they are
unused by default.

Make 'cfaMutation' option fit in 100 charwidth.

Add special behaviour if 'cfaMutation' is set in config checks.

o-petrov committed:
Add missed '\n'
